Implementation of GATs outlives lint

jackh726 opened the pull request:
See #87479 for background. Closes #87479

The basic premise of this lint/error is to require the user to write where clauses on a GAT when those bounds can be implied or proven from any function on the trait returning that GAT.

## Intuitive Explanation (Attempt) ##
Let's take this trait definition as an example:
```rust
trait Iterable {
    type Item<'x>;
    fn iter<'a>(&'a self) -> Self::Item<'a>;
}
```
Let's focus on the `iter` function. The first thing to realize is that we know that `Self: 'a` because of `&'a self`. If an impl wants `Self::Item` to contain any data with references, then those references must be derived from `&'a self`. Thus, they must live only as long as `'a`. Furthermore, because of the `Self: 'a` implied bound, they must live only as long as `Self`. Since it's `'a` is used in place of `'x`, it is reasonable to assume that any value of `Self::Item<'x>`, and thus `'x`, will only be able to live as long as `Self`. Therefore, we require this bound on `Item` in the trait.

As another example:
```rust
trait Deserializer<T> {
    type Out<'x>;
    fn deserialize<'a>(&self, input: &'a T) -> Self::Out<'a>;
}
```
The intuition is similar here, except rather than a `Self: 'a` implied bound, we have a `T: 'a` implied bound. Thus, the data on `Self::Out<'a>` is derived from `&'a T`, and thus it is reasonable to expect that the lifetime `'x` will always be less than `T`.

## Implementation Algorithm ##
* Given a GAT `<P0 as Trait<P1..Pi>>::G<Pi...Pn>` declared as `trait T<A1..Ai> for A0 { type G<Ai...An>; }` used in return type of one associated function `F`
* Given env `E` (including implied bounds) for `F`
* For each lifetime parameter `'a` in `P0...Pn`:
    * For each other type parameter `Pi != 'a` in `P0...Pn`: // FIXME: this include of lifetime parameters too
        * If `E => (P: 'a)`:
            * Require where clause `Ai: 'a`

## Follow-up questions ##
* What should we do when we don't pass params exactly?
For this example:
```rust
trait Des {
    type Out<'x, D>;
    fn des<'z, T>(&self, data: &'z Wrap<T>) -> Self::Out<'z, Wrap<T>>;
}
```
Should we be requiring a `D: 'x` clause? We pass `Wrap<T>` as `D` and `'z` as `'x`, and should be able to prove that `Wrap<T>: 'z`.


r? @nikomatsakis

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3927324720) [(plain)](https://github.com/rust-lang/rust/commit/29acb05a239c28ec2c4e8cd4bd4706d3b7ab40fd/checks/3927324720/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 12300/12307
.......
failures:

---- [ui] ui/generic-associated-types/self-outlives-lint.rs stdout ----


31    |                     help: add the required where clauses: `where U: 'y, T: 'x`
33 error: Missing required bounds on Out
-   --> $DIR/self-outlives-lint.rs:60:5
+   --> $DIR/self-outlives-lint.rs:59:5
35    |
35    |
36 LL |     type Out<'x, D>;


39    |                    help: add the required where clauses: `where D: 'x`
41 error: Missing required bounds on Out
-   --> $DIR/self-outlives-lint.rs:76:5
+   --> $DIR/self-outlives-lint.rs:75:5
43    |
43    |
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
44 LL |     type Out<'x, D>;


47    |                    help: add the required where clauses: `where D: 'x`
49 error: Missing required bounds on Out
-   --> $DIR/self-outlives-lint.rs:91:5
+   --> $DIR/self-outlives-lint.rs:90:5
51    |
51    |
52 LL |     type Out<'x, D>;


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/self-outlives-lint/self-outlives-lint.stderr
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/self-outlives-lint/self-outlives-lint.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args generic-associated-types/self-outlives-lint.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/self-outlives-lint" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/self-outlives-lint/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: Missing required bounds on Item
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:7:5
   |
LL |     type Item<'x>;
   |                  |
   |                  |
   |                  help: add the required where clauses: `where Self: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:23:5
   |
   |
LL |     type Out<'x>;
   |                 |
   |                 |
   |                 help: add the required where clauses: `where T: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:37:5
   |
   |
LL |     type Out<'x>;
   |                 |
   |                 |
   |                 help: add the required where clauses: `where T: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:44:5
   |
   |
LL |     type Out<'x, 'y>;
   |                     |
   |                     |
   |                     help: add the required where clauses: `where U: 'y, T: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:59:5
   |
   |
LL |     type Out<'x, D>;
   |                    |
   |                    |
   |                    help: add the required where clauses: `where D: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:75:5
   |
   |
LL |     type Out<'x, D>;
   |                    |
   |                    |
   |                    help: add the required where clauses: `where D: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:90:5
   |
   |
LL |     type Out<'x, D>;
   |                    |
   |                    |
   |                    help: add the required where clauses: `where D: 'x`
error: aborting due to 7 previous errors


------------------------------------------
---
test result: FAILED. 12189 passed; 1 failed; 117 ignored; 0 measured; 0 filtered out; finished in 127.01s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:13:29
```

</details>

jackh726 commented on src/test/ui/generic-associated-types/self-outlives-lint.stderr:
```diff
@@ -0,0 +1,98 @@
+LL |     type Bar<'b>;
+   |     ^^^^^^^^^^^^-
+   |                 |
+   |                 help: add the required where clauses: `where Self: 'b, Self: 'a`
```
This isn't right

jackh726 commented on src/test/ui/generic-associated-types/self-outlives-lint.stderr:
```diff
@@ -0,0 +1,98 @@
+LL |     type Bar<'b>;
+   |     ^^^^^^^^^^^^-
+   |                 |
+   |                 help: add the required where clauses: `where Self: 'b, Self: 'a`
```
Nor this.

jackh726 commented:
@nikomatsakis I've addressed the nits here. But have not yet reworked the diagnostic. I also added checking for region outlives region. I have not started working out the `Iterable` case and there tests where we have a lifetime on the *trait*, not the method could be cleaned up a bit (currently require `Self: 'a` and `Self: 'b`, when really we probably should pick one...maybe the trait parameter name)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4063632136?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ef5e31ac06716e2fd5c5f0dc0991b3760dc0c3ae/checks/4063632136/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.........................................................i.ii....................................... 12300/12370
......................................................................
failures:

---- [ui] ui/generic-associated-types/self-outlives-lint.rs stdout ----


28 LL |     type Out<'x, 'y>;
30    |                     |
30    |                     |
-    |                     help: add the required where clauses: `where U: 'y, T: 'x`
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
+    |                     help: add the required where clauses: `where T: 'x, U: 'y`
33 error: Missing required bounds on Out
34   --> $DIR/self-outlives-lint.rs:61:5



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/self-outlives-lint/self-outlives-lint.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args generic-associated-types/self-outlives-lint.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/self-outlives-lint" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/self-outlives-lint/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: Missing required bounds on Item
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:9:5
   |
LL |     type Item<'x>;
   |                  |
   |                  |
   |                  help: add the required where clauses: `where Self: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:25:5
   |
   |
LL |     type Out<'x>;
   |                 |
   |                 |
   |                 help: add the required where clauses: `where T: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:39:5
   |
   |
LL |     type Out<'x>;
   |                 |
   |                 |
   |                 help: add the required where clauses: `where T: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:46:5
   |
   |
LL |     type Out<'x, 'y>;
   |                     |
   |                     |
   |                     help: add the required where clauses: `where T: 'x, U: 'y`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:61:5
   |
   |
LL |     type Out<'x, D>;
   |                    |
   |                    |
   |                    help: add the required where clauses: `where D: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:77:5
   |
   |
LL |     type Out<'x, D>;
   |                    |
   |                    |
   |                    help: add the required where clauses: `where D: 'x`
error: Missing required bounds on Out
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:92:5
   |
   |
LL |     type Out<'x, D>;
   |                    |
   |                    |
   |                    help: add the required where clauses: `where D: 'x`
error: Missing required bounds on Bar
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:113:5
   |
   |
LL |     type Bar<'b>;
   |                 |
   |                 |
   |                 help: add the required where clauses: `where Self: 'b, Self: 'a`
error: Missing required bounds on Bar
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:120:5
   |
   |
LL |     type Bar<'b>;
   |                 |
   |                 |
   |                 help: add the required where clauses: `where Self: 'b, Self: 'a`
error: Missing required bounds on Bar
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:127:5
   |
   |
LL |     type Bar<'b>;
   |                 |
   |                 |
   |                 help: add the required where clauses: `where Self: 'b`
error: Missing required bounds on Iterator
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:141:5
   |
   |
LL |     type Iterator<'a>: Iterator<Item = Self::Item<'a>>;
   |                                                       |
   |                                                       |
   |                                                       help: add the required where clauses: `where Self: 'a`
error: Missing required bounds on Bar
  --> /checkout/src/test/ui/generic-associated-types/self-outlives-lint.rs:148:5
   |
   |
LL |     type Bar<'a, 'b>;
   |                     |
   |                     |
   |                     help: add the required where clauses: `where 'a: 'b`
error: aborting due to 12 previous errors


------------------------------------------
---
test result: FAILED. 12259 passed; 1 failed; 110 ignored; 0 measured; 0 filtered out; finished in 109.17s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:11:07
```

</details>

nikomatsakis commented on compiler/rustc_typeck/src/check/wfcheck.rs:
```diff
@@ -253,6 +259,337 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
+            regions: FxHashSet::default(),
+            types: FxHashSet::default(),
+        };
+        sig.output().visit_with(&mut visitor);
```
Total nit, but I'd prefer to package this up as a function, so that you write:

```rust
let (types, regions) = GATSubstCollector::visit(sig.output());
```

or something like that.

nikomatsakis reviewed (commented):
OK, this is looking pretty clean right now. r=me with the various comments addressed (or with FIXMEs created). I know it's not in the complete end state we want, but I'd also be ok to land it and iterate in tree.

jackh726 commented on compiler/rustc_typeck/src/check/wfcheck.rs:
```diff
@@ -253,6 +259,337 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
+        return;
+    }
+    let generics: &ty::Generics = tcx.generics_of(trait_item.def_id);
+    // If the current associated type doesn't have any (own) params, it's not a GAT
```
I added a FIXME.

jackh726 commented on compiler/rustc_typeck/src/check/wfcheck.rs:
```diff
@@ -253,6 +259,337 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
+            }
+        }
+
+        match clauses.as_mut() {
```
I added a comment with some example code. It does take the intersection, yes.

jackh726 commented on compiler/rustc_typeck/src/check/wfcheck.rs:
```diff
@@ -253,6 +259,337 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
+    })
+}
+
+fn region_known_to_outlive<'tcx>(
```
Yes, I left a FIXME.

jackh726 commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit b6edcbd7b5db718ff5e8ba229812b35f803ddd7a has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis b6edcbd7b5db718ff5e8ba229812b35f803ddd7a -->
<!-- homu: {"type":"Approved","sha":"b6edcbd7b5db718ff5e8ba229812b35f803ddd7a","approver":"nikomatsakis"} -->

bors commented:
:hourglass: Testing commit b6edcbd7b5db718ff5e8ba229812b35f803ddd7a with merge 9d39f6ab7dec5b8c6e8d9ce52a1f15b9e656c900...
<!-- homu: {"type":"BuildStarted","head_sha":"b6edcbd7b5db718ff5e8ba229812b35f803ddd7a","merge_sha":"9d39f6ab7dec5b8c6e8d9ce52a1f15b9e656c900"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4124736043?check_suite_focus=true)
Approved by: nikomatsakis
Pushing 9d39f6ab7dec5b8c6e8d9ce52a1f15b9e656c900 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nikomatsakis","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4124736043?check_suite_focus=true"},"merge_sha":"9d39f6ab7dec5b8c6e8d9ce52a1f15b9e656c900"} -->

rust-timer commented:
Finished benchmarking commit (9d39f6ab7dec5b8c6e8d9ce52a1f15b9e656c900): [comparison url](https://perf.rust-lang.org/compare.html?start=18cae2680fff914c95a633c468f09b614f385844&end=9d39f6ab7dec5b8c6e8d9ce52a1f15b9e656c900).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression

jackh726 commented:
@MingweiSamuel this is weird...maybe try to `cargo clean` and rebuild?

MingweiSamuel commented:
Didn't help and also happened in the fresh gh-actions ubuntu build. I will try to make a minimal example and submit a separate issue.

MingweiSamuel commented:
Turns out I had a dependency to an old git version in my workspace instead of the sibling project, issue is on my end between keyboard and chair :)
